Allow @sha256 digest for tags in FROM

When I try to use a digest tag in my Dockerfile FROM: such as "FROM: mytomcat@sha256:b61b1b823ec66a2875c44572885fa332990b3aa0055ed50abf7f94837c9f7925" I get this:

[ERROR] Failed to execute goal io.fabric8:docker-maven-plugin:0.15.14:build (docker-build) on project kestrel-docker: Execution docker-build of goal io.fabric8:docker-maven-plugin:0.15.14:build failed: Given repository name 'mytomcat@sha256:b61b1b823ec66a2875c44572885fa332990b3aa0055ed50abf7f94837c9f7925' is invalid:
[ERROR] \* image part 'mytomcat@sha256' doesn't match allowed pattern '(?:(?:[a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9-]_[a-zA-Z0-9])(?:(?:.(?:[a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9-]_[a-zA-Z0-9]))+)?(?::[0-9]+)?/)?[a-z0-9]+(?:(?:[._]|__|[-]_)[a-z0-9]+)_(?:/[a-z0-9]+(?:(?:[._]|__|[-]_)[a-z0-9]+)_)?'

I see various mentions in docker re: digest usage and our docker-compose.yml uses it just fine. Do the regex's need to be updated?

i'm still seeing this problem - even with docker-maven-plugin v 0.22.1 - ?

io.fabric8:docker-maven-plugin:0.22.1:build (build) on project portal-enterprise: Execution build of goal io.fabric8:docker-maven-plugin:0.22.1:build failed: Given Docker name 'docker.io/jetty@sha256:82d64b324a80ac32b3f6412f2d9fb9fa0b7ade76f748b9f0fd1a9bde416f7adc' is invalid:
[ERROR] * image part 'jetty@sha256' doesn't match allowed pattern '[a-z0-9]+(?:(?:(?:[._]|__|[-]*)[a-z0-9]+)+)?(?:(?:/[a-z0-9]+(?:(?:(?:[._]|__|[-]*)[a-z0-9]+)+)?)+)?'
[ERROR] See http://bit.ly/docker_image_fmt for more details

Dockerfile:

FROM docker.io/jetty@sha256:82d64b324a80ac32b3f6412f2d9fb9fa0b7ade76f748b9f0fd1a9bde416f7adc

pulling directly from the command line works, but i need to build my docker image using dockerfile

--- a/doc/changelog.md
+++ b/doc/changelog.md
@@ -6,6 +6,7 @@
   - Update to jnr-unixsocket version to 0.21 (#1089)
   - Add 'readOnly' option for docker:run cto mount container's root fs read-only (#1125)
   - Provide container properties to the wait configuration execution (#1111)
+  - Allow @sha256 digest for tags in FROM (#541)
 
 * **0.27.2** (2018-10-05)
 

--- a/src/main/java/io/fabric8/maven/docker/access/UrlBuilder.java
+++ b/src/main/java/io/fabric8/maven/docker/access/UrlBuilder.java
@@ -98,7 +98,7 @@ public final class UrlBuilder {
     public String pullImage(ImageName name, String registry) {
         return u("images/create")
                 .p("fromImage", name.getNameWithoutTag(registry))
-                .p("tag", name.getTag())
+                .p("tag", name.getDigest() != null ? name.getDigest() : name.getTag())
                 .build();
     }
 

--- a/src/main/java/io/fabric8/maven/docker/util/ImageName.java
+++ b/src/main/java/io/fabric8/maven/docker/util/ImageName.java
@@ -37,6 +37,9 @@ public class ImageName {
     // Tag name
     private String tag;
 
+    // Digest
+    private String digest;
+
     // User name
     private String user;
 
@@ -60,38 +63,26 @@ public class ImageName {
         if (fullName == null) {
             throw new NullPointerException("Image name must not be null");
         }
-        Pattern tagPattern = Pattern.compile("^(.+?)(?::([^:/]+))?$");
-        Matcher matcher = tagPattern.matcher(fullName);
-        if (!matcher.matches()) {
-            throw new IllegalArgumentException(fullName + " is not a proper image name ([registry/][repo][:port]");
-        }
-        tag = givenTag != null ? givenTag : matcher.group(2);
-        String rest = matcher.group(1);
 
-        String[] parts = rest.split("\\s*/\\s*");
-        if (parts.length == 1) {
-            registry = null;
-            user = null;
-            repository = parts[0];
-        } else if (parts.length >= 2) {
-            if (isRegistry(parts[0])) {
-                registry = parts[0];
-                if (parts.length > 2) {
-                    user = parts[1];
-                    repository = joinTail(parts);
-                } else {
-                    user = null;
-                    repository = parts[1];
-                }
-            } else {
-                registry = null;
-                user = parts[0];
-                repository = rest;
+        if(fullName.contains("@sha256")) { // Of it contains digest
+            String[] digestParts = fullName.split("@");
+            digest = digestParts[1];
+            parseComponentsBeforeTag(digestParts[0]);
+        } else {
+            digest = null;
+            Pattern tagPattern = Pattern.compile("^(.+?)(?::([^:/]+))?$");
+            Matcher matcher = tagPattern.matcher(fullName);
+            if (!matcher.matches()) {
+                throw new IllegalArgumentException(fullName + " is not a proper image name ([registry/][repo][:port]");
             }
-        }
+            tag = givenTag != null ? givenTag : matcher.group(2);
+            String rest = matcher.group(1);
 
-        if (tag == null) {
-            tag = "latest";
+            parseComponentsBeforeTag(rest);
+
+            if (tag == null) {
+                tag = "latest";
+            }
         }
 
         doValidate();
@@ -107,6 +98,10 @@ public class ImageName {
 
     public String getTag() {
         return tag;
+    }
+
+    public String getDigest() {
+        return digest;
     }
 
     public boolean hasRegistry() {
@@ -177,6 +172,9 @@ public class ImageName {
      * @return full name with original registry (if set) or optional registry (if not <code>null</code>).
      */
     public String getFullName(String optionalRegistry) {
+        if(digest != null) {
+            return getNameWithoutTag(optionalRegistry) + "@" + digest;
+        }
         return getNameWithoutTag(optionalRegistry) + ":" + tag;
     }
 
@@ -214,13 +212,14 @@ public class ImageName {
     // Validate parts and throw an IllegalArgumentException if a part is not valid
     private void doValidate() {
         List<String> errors = new ArrayList<>();
-        // Stripp of user from repository name
+        // Strip off user from repository name
         String image = user != null ? repository.substring(user.length() + 1) : repository;
         Object[] checks = new Object[] {
             "registry", DOMAIN_REGEXP, registry,
             "image", IMAGE_NAME_REGEXP, image,
             "user", NAME_COMP_REGEXP, user,
-            "tag", TAG_REGEXP, tag
+            "tag", TAG_REGEXP, tag,
+            "digest", DIGEST_REGEXP, digest
         };
         for (int i = 0; i < checks.length; i +=3) {
             String value = (String) checks[i + 2];
@@ -239,6 +238,30 @@ public class ImageName {
             }
             buf.append("See http://bit.ly/docker_image_fmt for more details");
             throw new IllegalArgumentException(buf.toString());
+        }
+    }
+
+    private void parseComponentsBeforeTag(String rest) {
+        String[] parts = rest.split("\\s*/\\s*");
+        if (parts.length == 1) {
+            registry = null;
+            user = null;
+            repository = parts[0];
+        } else if (parts.length >= 2) {
+            if (isRegistry(parts[0])) {
+                registry = parts[0];
+                if (parts.length > 2) {
+                    user = parts[1];
+                    repository = joinTail(parts);
+                } else {
+                    user = null;
+                    repository = parts[1];
+                }
+            } else {
+                registry = null;
+                user = parts[0];
+                repository = rest;
+            }
         }
     }
 
@@ -268,4 +291,6 @@ public class ImageName {
 
     // https://github.com/docker/docker/blob/04da4041757370fb6f85510c8977c5a18ddae380/vendor/github.com/docker/distribution/reference/regexp.go#L37
     private final Pattern TAG_REGEXP = Pattern.compile("^[\\w][\\w.-]{0,127}$");
+
+    private final Pattern DIGEST_REGEXP = Pattern.compile("^sha256:[a-z0-9]{32,}$");
 }

--- a/src/test/java/io/fabric8/maven/docker/util/ImageNameTest.java
+++ b/src/test/java/io/fabric8/maven/docker/util/ImageNameTest.java
@@ -125,6 +125,10 @@ public class ImageNameTest {
                 new ImageName("docker.jolokia.org/org/jolokia/jolokia_demo").getFullName("another.registry.org"));
         assertEquals("docker.jolokia.org/org/jolokia/jolokia_demo:latest",
                 new ImageName("docker.jolokia.org/org/jolokia/jolokia_demo").getFullName(null));
+        assertEquals("docker.jolokia.org/org/jolokia/jolokia_demo@sha256:2781907cc3ae9bb732076f14392128d4b84ff3ebb66379d268e563b10fbfb9da",
+                new ImageName("docker.jolokia.org/org/jolokia/jolokia_demo@sha256:2781907cc3ae9bb732076f14392128d4b84ff3ebb66379d268e563b10fbfb9da").getFullName(null));
+        assertEquals("docker.jolokia.org",
+                new ImageName("docker.jolokia.org/org/jolokia/jolokia_demo@sha256:2781907cc3ae9bb732076f14392128d4b84ff3ebb66379d268e563b10fbfb9da").getRegistry());
     }
 
 

